src/nnbench/context.py

```diff
@@ -183,3 +187,256 @@ def __call__(self) -> dict[str, Any]:
         result["memory_unit"] = self.memunit
         # TODO: Lacks CPU cache info, which requires a solution other than psutil.
         return {self.key: result}
+
+
+class Context:
+    def __init__(self, context: dict[str, Any] | ContextProvider | "Context" | None = None) -> None:
+        self._ctx_dict: dict[str, Any] = {}
+        if context:
+            self.merge(context, inplace=True)
+
+    def keys(self) -> KeysView[str]:
+        return self._ctx_dict.keys()
+
+    def values(self) -> ValuesView[Any]:
+        return self._ctx_dict.values()
+
+    def items(self) -> ItemsView[str, Any]:
+        return self._ctx_dict.items()
+
+    def merge(
+        self,
+        other: "Context" | dict[str, Any] | ContextProvider,
+        this_ctx: dict[str, Any] | None = None,
+        inplace: bool = True,
+    ) -> "Context":
+        """
+        Merge another dictionary, Context, or ContextProvider into this Context.
+
+        Parameters
+        ----------
+        other : Context | dict[str, Any] | ContextProvider
+            The other context to merge into this one.
+        this_ctx : dict[str, Any] | None, optional
+            The target dictionary to merge into. Takes this classes context if None.
+        inplace : bool, optional
+            If True, modifies the current context in place. Otherwise, returns a new merged Context.
+
+        Raises
+        ------
+        TypeError
+            If 'other' is not a Context, dict, or callable returning a dict, or if 'this_ctx'
+            is specified but not a dict.
+        ValueError
+            If 'other' is a callable that does not return a dictionary.
+
+        Returns
+        -------
+        Context
+            The current context after merging, if inplace is True. Otherwise, a new Context instance.
+        """
+        if isinstance(other, Context):
+            other = other._ctx_dict
+        elif callable(other):
+            result = other()
+            if not isinstance(result, dict):
+                raise ValueError("Provider did not return a dictionary.")
+            other = result
+        elif isinstance(other, dict):
+            pass
+        else:
+            raise TypeError(f"Unknown type for source, got {type(other)}")
```

Comment:
I think all of this would be dramatically easier if we operated on a list of items.

Proposition (as stated above): Instead of merge(), we could support an update() method. We lose the immutability of the left-hand side context, but I think that operation is much more common.